src/reaction.cpp

```diff
@@ -0,0 +1,277 @@
+#include "openmc/reaction.h"
+
+#include <string>
+#include <utility> // for move
+
+#include "openmc/constants.h"
+#include "openmc/hdf5_interface.h"
+#include "openmc/endf.h"
+#include "openmc/random_lcg.h"
+#include "openmc/secondary_uncorrelated.h"
+
+namespace openmc {
+
+//==============================================================================
+// Reaction implementation
+//==============================================================================
+
+Reaction::Reaction(hid_t group, const std::vector<int>& temperatures)
+{
+  read_attribute(group, "Q_value", q_value_);
+  read_attribute(group, "mt", mt_);
+  int tmp;
+  read_attribute(group, "center_of_mass", tmp);
+  scatter_in_cm_ = (tmp == 1);
+
+  // Checks if redudant attribute exists before loading
+  // (for compatibiltiy with legacy .h5 libraries)
+  if (attribute_exists(group, "redundant")) {
+    read_attribute(group, "redundant", tmp);
+    redundant_ = (tmp == 1);
+  } else {
+    redundant_ = false;
+  }
+
+  // Read cross section and threshold_idx data
+  for (auto t : temperatures) {
+    // Get group corresponding to temperature
+    std::string temp_str {std::to_string(t) + "K"};
+    hid_t temp_group = open_group(group, temp_str.c_str());
+    hid_t dset = open_dataset(temp_group, "xs");
+
+    // Get threshold index
+    TemperatureXS xs;
+    read_attribute(dset, "threshold_idx", xs.threshold);
+
+    // Read cross section values
+    read_dataset(dset, xs.value);
+    close_dataset(dset);
+    close_group(temp_group);
+
+    // create new entry in xs vector
+    xs_.push_back(std::move(xs));
+  }
+
+  // Read products
+  for (const auto& name : group_names(group)) {
+    if (name.rfind("product_", 0) == 0) {
+      hid_t pgroup = open_group(group, name.c_str());
+      products_.emplace_back(pgroup);
+      close_group(pgroup);
+    }
+  }
+
+  // <<<<<<<<<<<<<<<<<<<<<<<<<<<< REMOVE THIS <<<<<<<<<<<<<<<<<<<<<<<<<
+  // Before the secondary distribution refactor, when the angle/energy
+  // distribution was uncorrelated, no angle was actually sampled. With
+  // the refactor, an angle is always sampled for an uncorrelated
+  // distribution even when no angle distribution exists in the ACE file
+  // (isotropic is assumed). To preserve the RNG stream, we explicitly
+  // mark fission reactions so that we avoid the angle sampling.
+  if (is_fission(mt_)) {
+    for (auto& p : products_) {
+      if (p.particle_ == Particle::Type::neutron) {
+        for (auto& d : p.distribution_) {
+          auto d_ = dynamic_cast<UncorrelatedAngleEnergy*>(d.get());
+          if (d_) d_->fission() = true;
+        }
+      }
+    }
+  }
+  // <<<<<<<<<<<<<<<<<<<<<<<<<<<< REMOVE THIS <<<<<<<<<<<<<<<<<<<<<<<<<
+}
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+std::string reaction_name(int mt)
+{
+  if (mt == SCORE_FLUX) {
```

Comment:
consider splitting this if-else conditional into a hash_map lookup for string search cases.